packages/flutter/lib/src/widgets/will_pop_scope.dart

```diff
@@ -9,10 +9,53 @@ import 'routes.dart';
 /// Registers a callback to veto attempts by the user to dismiss the enclosing
 /// [ModalRoute].
 ///
+/// {@tool snippet --template=stateful_widget}
+///
+/// Whenever the back button is pressed, you will get a callback at [onWillPop],
+/// which returns a [Future]. If the [Future] returned is true, the screen is
+/// popped and if it is false, the screen is not popped.
+///
+/// ```dart
+/// @override
+/// Widget build(BuildContext) {
+///     return WillPopScope (
+///      onWillPop: _handleBackPress,
+///        child: MyWidget(),
+///     );
+/// }
+/// ```
+/// {@end-tool}
+///
+/// {@tool dartpad --template=stateless_widget_material}
+///
+/// ```dart
+/// Widget build(BuildContext context) {
+///   return Scaffold(
+///     appBar: AppBar(
+///       title: Text("WillPopScope demo"),
+///       leading: BackButton(),
+///     ),
+///     body: WillPopScope(
+///         child: Center(child: Text("Press back button in AppBar")),
+///         onWillPop: () async {
+///           ScaffoldMessenger.of(context).showSnackBar(
+///               SnackBar(
+///                 content: Text("Callback fired on pressing back button")));
```

Comment:
Rather than show a snackbar, it might be better to store a boolean on a stateful widget. You'll have to change this template to `stateful_widget_material` and add some way to toggle:

```dart
class _MyStatefulWidgetState extends State<Page2> {
  bool shouldPop = true;

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        return shouldPop;
      },
      child: Center(
        child: OutlinedButton(
          child: Text('shouldPop: $shouldPop'),
          onPressed: () {
            setState(() {
              shouldPop = !shouldPop;
            });
          },
        ),
      ),
    );
  }
}
```

Reply:
I am thinking that too make things even more clear, I will add two pages. I will add the toggle in the second page, where the page will pop depending on what the user toggles